### PR TITLE
Possible fix for #149

### PR DESCRIPTION
**Pull request by smarttowel**

### CODE REVIEW

**Comment by TheAssassin** [comment]
I think the reason the rpath was not touched if set already in linuxdeployqt is that too many assumptions were made while this was implemented. The only reason linuxdeploy does this this way is because linuxdeployqt does, too. I do not think it makes sense to just ignore the `rpath` altogether.

We should instead try to add (prepend/append) `$ORIGIN` to an existing rpath if it doesn't contain the value already. Could you please update your PR to implement such a behavior? I'm not sure prepending makes a lot of sense, I'd rather append, I guess.

**Comment by smarttowel** [comment]
Main reason of this PR is to fix gstreamer plugin that is broken now. It's happened, because linuxdeploy override rpath that gstreamer-plugin set. And, maybe, it's a fix for #149.

> We should instead try to add (prepend/append) `$ORIGIN` to an existing rpath if it doesn't contain the value already. Could you please update your PR to implement such a behavior? I'm not sure prepending makes a lot of sense, I'd rather append, I guess.

Done.

File: include/linuxdeploy/util/misc.h [hunk 1]
```diff
@@ -56,6 +56,17 @@ namespace linuxdeploy {
                 return split(s, '\n');
             }
 
+            static std::string join(const std::vector<std::string> &strings, const std::string &delimiter)
```
Comment (TheAssassin) [inline include/linuxdeploy/util/misc.h:59]: :+1:

**Review by TheAssassin — CHANGES_REQUESTED** [review]

File: src/core/appdir.cpp [hunk 2]
```diff
@@ -834,7 +834,15 @@ namespace linuxdeploy {
                     if (!d->deployElfDependencies(sharedLibrary))
                         return false;
 
-                    d->setElfRPathOperations[sharedLibrary] = "$ORIGIN";
+                    auto rpath = elf_file::ElfFile(sharedLibrary).getRPath();
```
[the remark this answers is not in the document]
Comment (smarttowel) [reply src/core/appdir.cpp]: Done

File: src/core/appdir.cpp [hunk 3]
```diff
@@ -834,7 +834,15 @@ namespace linuxdeploy {
...
 
-                    d->setElfRPathOperations[sharedLibrary] = "$ORIGIN";
+                    const auto rpath = elf_file::ElfFile(sharedLibrary).getRPath();
+                    auto rpathList = util::split(rpath, ':');
+                    if (std::find(rpathList.begin(), rpathList.end(), "$ORIGIN") == rpathList.end())
+                    {
+                        rpathList.push_back("$ORIGIN");
+                        d->setElfRPathOperations[sharedLibrary] = util::join(rpathList, ":");
+                    } else {
+                        d->setElfRPathOperations[sharedLibrary] = rpath;
```
Comment (TheAssassin) [inline src/core/appdir.cpp:844]: Just an idea (I don't insist on it), couldn't we spare this one and just call `d->setElfRPathOperations[sharedLibrary] = util::join(rpathList, ":");` once after the `if`?

File: src/core/appdir.cpp [hunk 4]
```diff
@@ -834,7 +834,15 @@ namespace linuxdeploy {
                     if (!d->deployElfDependencies(sharedLibrary))
                         return false;
 
-                    d->setElfRPathOperations[sharedLibrary] = "$ORIGIN";
+                    const auto rpath = elf_file::ElfFile(sharedLibrary).getRPath();
+                    auto rpathList = util::split(rpath, ':');
+                    if (std::find(rpathList.begin(), rpathList.end(), "$ORIGIN") == rpathList.end())
+                    {
```
Comment (TheAssassin) [inline src/core/appdir.cpp:840]:
```suggestion
                    if (std::find(rpathList.begin(), rpathList.end(), "$ORIGIN") == rpathList.end()) {
```
Code style: `{` must be on the same line the `if` is on.

**Review by TheAssassin — CHANGES_REQUESTED** [review]

File: src/core/appdir.cpp (hunk 3 above)
Comment (smarttowel) [reply src/core/appdir.cpp:844]: We can, but it's call `join(rpathList)` even than rpath not changed.

Comment (TheAssassin) [reply src/core/appdir.cpp:843]: It won't make a difference, I guess, just saves two lines. Let's keep it like this. Thanks!

File: include/linuxdeploy/util/misc.h [hunk 5]
```diff
@@ -56,6 +56,17 @@ namespace linuxdeploy {
...
+            {
```
Comment (TheAssassin) [inline include/linuxdeploy/util/misc.h:60]:
```suggestion
            static std::string join(const std::vector<std::string> &strings, const std::string &delimiter) {
```

File: include/linuxdeploy/util/misc.h [hunk 6]
```diff
@@ -56,6 +56,17 @@ namespace linuxdeploy {
...
+                std::string result;
+                for (size_t i = 0; i < strings.size(); i++) {
+                    result += strings[i];
+                    if (i != strings.size() - 1)
+                        result += delimiter;
+                }
```
Comment (TheAssassin) [inline include/linuxdeploy/util/misc.h:66]:
```suggestion

                    if (i != strings.size() - 1) {
                        result += delimiter;
                    }
                }
```

**Review by TheAssassin — APPROVED** [review]

**Comment by TheAssassin** [comment]
Thanks!

**Pull request merged**